File: asme/ops/serializers/attachments.py
"""JSON shape for attachments.

``download_url`` is a relative URL carrying a short-lived token bound to the
attachment id and the viewer, minted fresh on every serialisation.
"""

from __future__ import annotations

from flask import url_for

from asme.ops import storage
from asme.ops.serializers import iso, uid, user_ref


def download_url(row, user_id: int | None) -> str:
    token = storage.sign_download(row.id, user_id)
    return url_for("ops_api.download_attachment", attachment_id=uid(row.id), token=token)


def attachment(row, *, user_id: int | None) -> dict:
    return {
        "id": uid(row.id),
        "entity_type": row.entity_type,
        "entity_id": uid(row.entity_id),
        "original_name": row.original_name,
        "content_type": row.content_type,
        "size_bytes": int(row.size_bytes or 0),
        "is_image": bool(row.is_image),
        "uploaded_by": user_ref(row.uploaded_by),
        "download_url": download_url(row, user_id),
        "created_at": iso(row.created_at),
    }

File: apps/ops-web/src/api/contracts/attachments.ts
import { z } from 'zod'
import { listOf, UserRef } from './common'

/** Shape from asme/ops/serializers/attachments.py; `download_url` carries a short-lived signed token. */
export const Attachment = z.looseObject({
  id: z.string(),
  entity_type: z.string(),
  entity_id: z.string(),
  original_name: z.string(),
  content_type: z.string().nullable().optional(),
  size_bytes: z.number().default(0),
  is_image: z.boolean().optional(),
  uploaded_by: UserRef.nullable(),
  download_url: z.string(),
  created_at: z.string(),
})
export type Attachment = z.infer<typeof Attachment>

export const AttachmentList = listOf(Attachment)
export type AttachmentList = z.infer<typeof AttachmentList>

/** URL segment accepted by `/:entity/:id/attachments`. */
export type AttachmentEntity = 'work-orders' | 'projects' | 'assets'

export function isImageAttachment(attachment: Pick<Attachment, 'is_image' | 'content_type'>): boolean {
  return attachment.is_image ?? Boolean(attachment.content_type?.startsWith('image/'))
}

export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(bytes < 10 * 1024 ? 1 : 0)} KB`
  return `${(bytes / 1024 / 1024).toFixed(1)} MB`
}
